src/kboot_gpu.c

```diff
@@ -674,5 +687,54 @@ int dt_set_gpu(void *dt)
     if (firmware_set_fdt(dt, gpu, "apple,firmware-compat", compat))
         return -1;
 
+    size_t data_a_size, data_b_size, globals_size;
+    if (rust_gpu_initdata_size(compat->num[0], compat->num[1], &data_a_size, &data_b_size,
+                               &globals_size) == -1)
+        return -1;
+
+    void *data_a = calloc(1, data_a_size);
+    void *data_b = calloc(1, data_b_size);
+    void *globals = calloc(1, globals_size);
+    size_t n_perf_states_cs = 0, n_perf_states_afr = 0;
+    const struct aux_perf_state *pstate_cs_raw = NULL, *pstate_afr_raw = NULL;
+    u64 uat_ttb_base;
+    ADT_GETPROP(adt, sgx, "gpu-region-base", &uat_ttb_base);
+    u64 gpu_base;
+
+    int adt_sgx_path[8];
+    if (adt_path_offset_trace(adt, "/arm-io/sgx", adt_sgx_path) < 0)
+        bail("ADT: GPU: Failed to get sgx\n");
+
+    if (adt_get_reg(adt, adt_sgx_path, "reg", 0, &gpu_base, NULL) < 0)
+        bail("ADT: GPU: Failed to get sgx reg 0\n");
+
+    u32 base_pstate;
+    if (ADT_GETPROP(adt, sgx, "gpu-perf-base-pstate", &base_pstate) < 0)
+        bail("ADT: GPU: missing gpu-perf-base-pstate\n");
+
+    u32 id_version = read32(gpu_base + 0xd04000);
+    u32 num_cores = read32(gpu_base + 0xd04010) & 0xff;
+    u32 gpu_rev_id = (id_version >> 8) & 0xff;
+    if (has_cs_afr) {
+        n_perf_states_cs = perf_states_cs->count;
+        n_perf_states_afr = perf_states_afr->count;
+        pstate_cs_raw = perf_states_cs->states;
+        pstate_afr_raw = perf_states_afr->states;
+    }
+    if (rust_fill_gpu_initdata(
+            perf_state_table_count, perf_state_count, perf_states, max_pwr, core_leak, sram_leak,
+            cs_leak, afr_leak, n_perf_states_cs, pstate_cs_raw, n_perf_states_afr, pstate_afr_raw,
+            adt_is_compatible(adt, 0, "J180dAP"), base_pstate, uat_ttb_base, num_cores,
```

Comment:
handling `J180d` specially should not be necessary. This seems to be an bug in our devicetrees. The J475d ADT uses the exact same properties as specified in `t6022-j180d.dts`